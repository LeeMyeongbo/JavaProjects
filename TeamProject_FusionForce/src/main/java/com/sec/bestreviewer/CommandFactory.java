package com.sec.bestreviewer;

import com.sec.bestreviewer.command.*;
import com.sec.bestreviewer.store.Employee;
import com.sec.bestreviewer.store.EmployeeStore;
import com.sec.bestreviewer.util.OptionParser;
import com.sec.bestreviewer.util.Pair;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CommandFactory {

    public static final String CMD_ADD = "ADD";
    public static final String CMD_DEL = "DEL";
    public static final String CMD_SCH = "SCH";
    public static final String CMD_CNT = "CNT";
    public static final String CMD_MOD = "MOD";

    private static final String EMPLOYEE_NUMBER = "employeeNum";
    private static final String NAME = "name";
    private static final String CAREER_LEVEL = "cl";
    private static final String PHONE_NUMBER = "phoneNum";
    private static final String BIRTHDAY = "birthday";
    private static final String CERTI = "certi";

    private static final Map<String, String> fieldMap = new HashMap<>();

    static {
        fieldMap.put(EMPLOYEE_NUMBER, EmployeeStore.FIELD_EMPLOYEE_NUMBER);
        fieldMap.put(NAME, EmployeeStore.FIELD_NAME);
        fieldMap.put(CAREER_LEVEL, EmployeeStore.FIELD_CAREER_LEVEL);
        fieldMap.put(PHONE_NUMBER, EmployeeStore.FIELD_PHONE_NUMBER);
        fieldMap.put(BIRTHDAY, EmployeeStore.FIELD_BIRTH_DAY);
        fieldMap.put(CERTI, EmployeeStore.FIELD_CERTI);
    }

    private CommandFactory() {
    }

    public static Command buildCommand(String cmd, List<String> options, List<String> params)
            throws IllegalArgumentException {
        final OptionParser optionParser = new OptionParser(options);

        return switch (cmd) {
            case CMD_ADD -> {
                final Employee employee = new Employee(params.get(0), params.get(1), params.get(2),
                    params.get(3), params.get(4), params.get(5));
                yield new AddCommand(optionParser, employee);
            }
            case CMD_DEL -> new DeleteCommand(optionParser, getConditionMapFromParams(params));
            case CMD_SCH -> new SearchCommand(optionParser, getConditionMapFromParams(params));
            case CMD_CNT -> new CountCommand();
            case CMD_MOD -> new ModifyCommand(optionParser, getConditionMapFromParams(params));
            default -> throw new IllegalArgumentException("Wrong command");
        };
    }

    private static List<Pair<String, String>> getConditionMapFromParams(List<String> params) {
        List<Pair<String, String>> conditionPairs = new ArrayList<>();
        for (int i = 0; i < params.size(); i += 2) {
            final String fieldName = fieldMap.get(params.get(i));
            if (fieldName == null) {
                throw new IllegalArgumentException("Wrong field: " + params.get(i));
            }
            conditionPairs.add(Pair.create(fieldName, params.get(i + 1)));
        }
        return conditionPairs;
     }
}
