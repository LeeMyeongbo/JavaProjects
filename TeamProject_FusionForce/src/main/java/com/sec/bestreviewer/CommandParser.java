package com.sec.bestreviewer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CommandParser {

    private final static int MIN_TOKENS_NUM = 5;
    private final static String SPACE = " ";
    private final static String EMPTY = "";

    public TokenGroup parse(String line) throws IllegalArgumentException {
        String[] r = line.split(",", -1);

        if (r.length < MIN_TOKENS_NUM) {
            throw new IllegalArgumentException("wrong command format");
        }

        String type = r[0];
        List<String> options = new ArrayList<>();
        List<String> params = new ArrayList<>();

        if (r.length > 8 && (r[6].equals("-a") || r[6].equals("-o"))) {
            options.addAll(Arrays.asList(Arrays.copyOfRange(r, 1, 4)));
            options.addAll(Arrays.asList(Arrays.copyOfRange(r, 6, 9)));
            params.addAll(Arrays.asList(Arrays.copyOfRange(r, 4, 6)));
            params.addAll(Arrays.asList(Arrays.copyOfRange(r, 9, r.length)));
        } else {
            options.addAll(Arrays.asList(Arrays.copyOfRange(r, 1, 4)));
            params.addAll(Arrays.asList(Arrays.copyOfRange(r, 4, r.length)));
        }

        return new TokenGroup(type, options, getValidList(params));
    }

    private List<String> getValidList(List<String> list) {
        List<String> validList = new ArrayList<>();

        list.forEach(i -> {
            if (!EMPTY.equals(i) && !SPACE.equals(i)) {
                validList.add(i);
            }
        });

        return validList;
    }
}
