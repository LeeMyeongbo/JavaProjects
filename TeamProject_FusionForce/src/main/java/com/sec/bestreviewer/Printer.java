package com.sec.bestreviewer;

import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Printer {

    private final String outputFilName;
    private final StringBuilder stdStringBuilder;

    public Printer(String outputFilename) {
        this(outputFilename, new StringBuilder());
    }

    public Printer(String outputFilename, StringBuilder stdStringBuilder) {
        this.outputFilName = outputFilename;
        this.stdStringBuilder = stdStringBuilder;
    }

    public void add(List<String> lines) {
        for (String line : lines) {
            stdStringBuilder.append(line).append("\n");
        }
    }

    public void printOutputFile() {
        try (OutputStream output = Files.newOutputStream(Paths.get(outputFilName))) {
            output.write(stdStringBuilder.toString().getBytes());
        } catch (Exception e) {
            e.getStackTrace();
        }
    }
}
