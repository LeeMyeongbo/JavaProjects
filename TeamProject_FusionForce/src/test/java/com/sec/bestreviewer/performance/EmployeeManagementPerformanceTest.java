package com.sec.bestreviewer.performance;

import com.sec.bestreviewer.EmployeeManagement;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.text.DecimalFormat;

public class EmployeeManagementPerformanceTest {

    @Test
    public void deletePerformanceTest() {
        final String outputFileName = "./src/test/java/com/sec/bestreviewer/performance/delete_performance_test_output.txt";
        final String inputFileName = "./src/test/java/com/sec/bestreviewer/performance/delete_performance_test_input.txt";

        String[] args = {inputFileName, outputFileName};
        File outputFile = new File(outputFileName);

        EmployeeManagement employeeManagement = new EmployeeManagement();

        long startTime = System.nanoTime();
        employeeManagement.run(args);
        long endTime = System.nanoTime();

        double durationSec = (double)(endTime - startTime) / 1000000000.0; // 나노초를 초로 변환
        DecimalFormat df = new DecimalFormat("#.###"); // 소수점 세 자리까지 출력
        System.out.println("Execution time: " + df.format(durationSec) + "s");

        Approvals.verify(outputFile);

    }

    @Test
    public void searchPerformanceTest() {
        final String outputFileName = "./src/test/java/com/sec/bestreviewer/performance/search_performance_test_output.txt";
        final String inputFileName = "./src/test/java/com/sec/bestreviewer/performance/search_performance_test_input.txt";

        String[] args = {inputFileName, outputFileName};
        File outputFile = new File(outputFileName);

        EmployeeManagement employeeManagement = new EmployeeManagement();

        long startTime = System.nanoTime();
        employeeManagement.run(args);
        long endTime = System.nanoTime();

        double durationSec = (double)(endTime - startTime) / 1000000000.0; // 나노초를 초로 변환
        DecimalFormat df = new DecimalFormat("#.###"); // 소수점 세 자리까지 출력
        System.out.println("Execution time: " + df.format(durationSec) + "s");

        Approvals.verify(outputFile);
    }

    @Test
    public void searchThirdOptionPerformanceTest() {
        final String outputFileName = "./src/test/java/com/sec/bestreviewer/performance/search_third_option_performance_test_output.txt";
        final String inputFileName = "./src/test/java/com/sec/bestreviewer/performance/search_third_option_performance_test_input.txt";

        String[] args = {inputFileName, outputFileName};
        File outputFile = new File(outputFileName);

        EmployeeManagement employeeManagement = new EmployeeManagement();

        long startTime = System.nanoTime();
        employeeManagement.run(args);
        long endTime = System.nanoTime();

        double durationSec = (double)(endTime - startTime) / 1000000000.0; // 나노초를 초로 변환
        DecimalFormat df = new DecimalFormat("#.###"); // 소수점 세 자리까지 출력
        System.out.println("Execution time: " + df.format(durationSec) + "s");

        Approvals.verify(outputFile);
    }

    @Test
    public void modifyPerformanceTest() {
        final String outputFileName = "./src/test/java/com/sec/bestreviewer/performance/modify_performance_test_output.txt";
        final String inputFileName = "./src/test/java/com/sec/bestreviewer/performance/modify_performance_test_input.txt";

        String[] args = {inputFileName, outputFileName};
        File outputFile = new File(outputFileName);

        EmployeeManagement employeeManagement = new EmployeeManagement();

        long startTime = System.nanoTime();
        employeeManagement.run(args);
        long endTime = System.nanoTime();

        double durationSec = (double)(endTime - startTime) / 1000000000.0; // 나노초를 초로 변환
        DecimalFormat df = new DecimalFormat("#.###"); // 소수점 세 자리까지 출력
        System.out.println("Execution time: " + df.format(durationSec) + "s");


        Approvals.verify(outputFile);
    }
}
