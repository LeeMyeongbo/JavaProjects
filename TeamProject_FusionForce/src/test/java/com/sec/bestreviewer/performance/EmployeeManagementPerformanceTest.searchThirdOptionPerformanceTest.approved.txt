SCH,90000029,RVYZGH USC,CL2,010-9048-1897,20181027,EX
SCH,90000090,RDVQHY JJW,CL1,010-9546-5033,20180521,PRO
SCH,90000110,CLGYUA KHU,CL3,010-7869-4474,20171104,PRO
SCH,90000112,QOUBHJ XPN,CL2,010-9416-6147,20171121,ADV
SCH,90000438,JTNKIH UXM,CL2,010-7939-7646,20190810,ADV
SCH,90000007,EZHYSW GHG,CL2,010-8703-1789,19970112,EX
SCH,90000009,FKMNKA XPD,CL2,010-1586-1688,20060112,PRO
SCH,90000012,DYXMLM DJE,CL1,010-0070-2148,19790228,ADV
SCH,90000014,LVACFQ MMR,CL2,010-1456-1370,19770719,PRO
SCH,90000025,RNNTJF KVI,CL4,010-6824-0327,20060212,ADV
SCH,90000015,TERSJY WHR,CL4,010-1972-3784,20160719,EX
SCH,90000016,WTLFQD AZK,CL3,010-5182-6126,20180328,EX
SCH,90000029,RVYZGH USC,CL2,010-9048-1897,20181027,EX
SCH,90000033,BDUGKS OGU,CL1,010-6113-8174,20191219,EX
SCH,90000065,ISZTUW KEQ,CL2,010-0710-6056,20190803,EX
